### PR TITLE
Swift: detect hardcoded encryption keys

### PR DESCRIPTION
**Pull request by karimhamdanali**
Hardcoded keys should not be used for creating encryption ciphers. Data encrypted using hardcoded keys are more vulnerable to the possibility of recovering them.

The rule currently supports all ciphers that the CryptoSwift API provides, but we can always extend it further if more cophers are added.

I'd appreciate a review of the query itself, the accompanying tests, and the associated documentation.

### CODE REVIEW

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * An `Expr` that is used to initialize a key.
+ */
+abstract class KeySource extends Expr { }
+
+/**
+ * The creation of a literal byte array.
+ */
+class ByteArrayLiteralSource extends KeySource {
+  ByteArrayLiteralSource() { this = any(ArrayExpr arr | arr.getType().toString() = "Array<UInt8>") }
```
Comment (MathiasVP) [inline swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:26]: I _think_ `toString` does the same as `getName` here, right? In that case, I'd prefer using `getName` here (since we like to be able to mess with the output of `toString` without fear of breaking query logic).

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+/**
+ * Any string literal as a key source.
+ */
+class StringLiteralSource extends KeySource {
+  StringLiteralSource() { this instanceof StringLiteralExpr }
+}
```
Comment (MathiasVP) [inline swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:34]: If you want to be fancy:
```suggestion
class StringLiteralSource extends KeySource instanceof StringLiteralExpr { }
```
You can read about this use of `instanceof` [here](https://codeql.github.com/docs/ql-language-reference/types/#non-extending-subtypes).

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * A class for all ways to set a key.
+ */
+class EncryptionKeySink extends Expr {
+  EncryptionKeySink() {
+    // `key` arg in `init` is a sink
+    exists(ClassDecl c, AbstractFunctionDecl f, CallExpr call |
```
Comment (MathiasVP) [inline swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:42]: Should this `ClassDecl` be a `ClassOrStructDecl`? Or do we know that they're all actually `class`es?

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @name Hard-coded encryption key
+ * @description Using hardcoded keys for encryption is not secure, because potential attacker can easiy guess them.
```
Comment (MathiasVP) [inline swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:3]: Did you intend to write `because a potential attacker`? Or maybe `because potential attackers`?

**Review by MathiasVP — COMMENTED** [review]
This looks great so far!

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql (hunk 1 above)
Comment (karimhamdanali) [reply swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:26]: Agreed, and it does behave the same way. Confirmed by re-running the tests, so I've changed it to `getName` instead.

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql (hunk 2 above)
Comment (karimhamdanali) [reply swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:34]: Neat! Changed it to the less verbose syntax that you suggested.

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql (hunk 3 above)
Comment (karimhamdanali) [reply swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:42]: They are indeed all classes.

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql (hunk 4 above)
Comment (karimhamdanali) [reply swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql:3]: I intended the latter. Fixed it now.

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.qhelp [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>Hardcoded keys should not be used for creating encryption ciphers. Data encrypted using hardcoded keys are more vulnerable to the possibility of recovering them.</p>
+  </overview>
+
+  <recommendation>
+    <p>Use a randomly generated key material to initialize the encryption cipher.</p>
```
Comment (geoffw0) [inline swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.qhelp:10]:
```suggestion
    <p>Use a randomly generated key to initialize the encryption cipher.</p>
```
or
```
    <p>Use randomly generated key material to initialize the encryption cipher.</p>
```

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	_ = try AES(key: keyString, iv: ivString)
+
+
+
+	// GOOD: Using randomly generated keys for encryption
+	let key = (0..<10).map({ _ in UInt8.random(in: 0...UInt8.max) })
+	let keyString = String(cString: key)
+	let ivString = getRandomIV()
+	_ = try Blowfish(key: key, blockMode: CBC(), padding: padding)
+	_ = try Blowfish(key: keyString, iv: ivString)
```
Comment (geoffw0) [inline swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.swift:19]: I think using a different encryption algorithm in the `GOOD` cases distracts from the point, which is that you're using a randomly generated key in this version.  Perhaps use AES everywhere, or AES and Blowfish once each equally in the `GOOD` and `BAD` block?

File: swift/ql/test/query-tests/Security/CWE-321/test.swift [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+// Helper functions
+func getConstantString() -> String {
+  "this string is constant"
+}
+
+func getConstantArray() -> Array<UInt8> {
+	[UInt8](getConstantString().utf8)
+}
```
Comment (geoffw0) [inline swift/ql/test/query-tests/Security/CWE-321/test.swift:81]: `getConstantArray` is currently unused in the test.

**Review by geoffw0 — COMMENTED** [review]
Looks great, a few minor suggestions.

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.swift (hunk 6 above)
Comment (karimhamdanali) [reply swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.swift:19]: You're right. I've added both algorithms to show that this vulnerability is not specific to just AES/Blowfish.

File: swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.qhelp (hunk 5 above)
Comment (karimhamdanali) [reply swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.qhelp:10]: Fixed.

File: swift/ql/test/query-tests/Security/CWE-321/test.swift (hunk 7 above)
Comment (karimhamdanali) [reply swift/ql/test/query-tests/Security/CWE-321/test.swift:81]: Right. I was planning on using this function for a corner case that deals with constant propagation. It's a true negative though, so I'd have to properly document it in the test.

Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-321/test.swift:81]: Its fine to have test cases we don't catch or don't yet catch.  Just mark them with `// [NOT DETECTED]` or similar.

**Review by stevecat — APPROVED** [review]
:wave: Reviewing on behalf of the docs team. This looks great! ✨

**Review by geoffw0 — APPROVED** [review]
LGTM. 🎉
